std/unicode.zig

```diff
@@ -12,6 +12,47 @@ pub fn utf8ByteSequenceLength(first_byte: u8) !u3 {
     return error.Utf8InvalidStartByte;
 }
 
+pub fn utf8Encode(c: u32, out: &[]const u8) !void {
+    if (c < 0x80) {
+        // Is made up of one byte
+        // Can just add a '0' and then the code point
+        // Thus can just output 'c'
+        var result: [1]u8 = undefined;
```

Comment:
@andrewrk, @Hejsil 
This is an idiom that I've done a lot since returning arrays from functions ain't pretty due to the arrays being allocated within the function stack, is there a better way to get an array output, and furthermore is there a better way to assign a slice to a new slice size without doing this 'trick'.

In most languages there is a syntax for creating a new array on the stack and assigning its value to something else, currently this requires a hidden memcpy presumably; which I think we could optimise away.

Reply:
This is returning stack-allocated data from within the function to the caller and isn't valid.

Consider the following example:

```
const std = @import("std");
const warn = std.debug.warn;

fn example(out: &[]const u8, s: u8) void {
    var a: [4]u8 = undefined;
    a[0] = s;
    a[1] = s + 1;
    a[2] = s + 2;
    a[3] = s + 3;
    *out = a;
}

pub fn main() void {
    var a: []const u8 = undefined;
    example(&a, 0);

    warn("a: "); for (a) |e| { warn("{} ", e); } warn("\n");

    var b: []const u8 = undefined;
    example(&b, 1);

    warn("b: "); for (b) |e| { warn("{} ", e); } warn("\n");
    warn("a: "); for (a) |e| { warn("{} ", e); } warn("\n");
}
```

Gives the output:

```
$ zig run example.zig
a: 0 1 2 3 
b: 1 2 3 4 
a: 1 2 3 4
```

That is the best case that is happening, there are no guarantees about the data returned since everything within the function call is out of scope on exit unless it was dynamically allocated.

I would suggest instead taking a raw slice and simply asserting that the user has provided a sufficiently sized buffer. This allows the user to use their own buffer which they may want and provides the most flexibility.

Fixing up that previous example:

```
const std = @import("std");
const warn = std.debug.warn;

fn example(out: []u8, s: u8) void {
    std.debug.assert(out.len >= 4);
    out[0] = s;
    out[1] = s + 1;
    out[2] = s + 2;
    out[3] = s + 3;
}

pub fn main() void {
    var a: [4]u8 = undefined;
    example(a[0..], 0);

    warn("a: "); for (a) |e| { warn("{} ", e); } warn("\n");

    var b: [4]u8 = undefined;
    example(b[0..], 1);

    warn("b: "); for (b) |e| { warn("{} ", e); } warn("\n");
    warn("a: "); for (a) |e| { warn("{} ", e); } warn("\n");
}
```

If you really wanted to return fixed data you could create a new struct and wrap the array. This requires bounding the array size to the maximum needed. This is less flexible and I wouldn't implement it this way.

```
const std = @import("std");
const warn = std.debug.warn;

const Encoded = struct { out: [4]u8 };

fn example(s: u8) Encoded {
    var a: Encoded = undefined;
    a.out[0] = s;
    a.out[1] = s + 1;
    a.out[2] = s + 2;
    a.out[3] = s + 3;
    return a;
}

pub fn main() void {
    const a = example(0);

    warn("a: "); for (a.out) |e| { warn("{} ", e); } warn("\n");

    const b = example(1);

    warn("b: "); for (b.out) |e| { warn("{} ", e); } warn("\n");
    warn("a: "); for (a.out) |e| { warn("{} ", e); } warn("\n");
}
```

Reply:
To return the length I'd keep it simple and just take an integer pointer here.

Reply:
The only problem with it is that this is used for encoding bytes back to a byte[] which means the size isn't known prior for example, this means that it gets a bit finicky.  The way I think may be better for this unknown size thing could either be requiring them to run a function to get how many bytes the code point will be OR;
```C
fn example(s: [4]const u8) []const u8 {
    // ... do stuff with 's'
    return s[0..sizeWanted];
}
```
Perhaps?  Or just result to a full buffer based system and return the length of that byte?  Giving them the opportunity to either slice it or leave it?  Tricky problem this one is.

Reply:
I agree, I don't like the struct way; it feels icky.

Reply:
I would just require the caller to provide a buffer of at least 4 bytes. They don't know the size in advance so they need to provide the worst case scenario. Even if they did know the size, that would be a runtime value and they couldn't allocate exactly that size unless they use an `Allocator` for that, which exceeds more than 4-bytes of overhead in bookkeeping.

Reply:
Fair enough okay, and style was returning the length actually used is preferred to returning the slice?

Reply:
You could do either, largely depends on the expected purpose but they both will provide the same information.

Reply:
You could do this, and get a static guarantee that the buffer is a certain size:
```
fn example(out: &[4]u8, s: u8) void {
    (*out)[0] = s;
    (*out)[1] = s + 1;
    (*out)[2] = s + 2;
    (*out)[3] = s + 3;
}
```

Related: #863

Reply:
Oooh I do like that too.

Reply:
One issue with requiring an array of a specific size is that you are forcing the caller to use an array of this type. The caller will need to memcpy the bytes if they want to move the bytes into any other sized array. The static guarantees are nice, but if the data is copied anywhere else you will be using a slice interface anyway so will simply be performing the implied check later down the track.

Also, does anywhere else in std take a fixed array argument?

Reply:
I suppose the main point I'm trying to say is that since a slice is a view into a backing array so in my view it doesn't make sense to limit the output type to an array of a specific size, when you could use a slice and take an arbitrary sized array.

Your proposal looks nice and would be valuable here. For now, I'd stick with taking a slice and if down the track we get better guarantees we can update any of these occurences.